### ISSUE TITLE
Core Data tracking reports errors when none actually exist

### ISSUE THREAD

**Issue by bjhomer** [getsentry/sentry-cocoa#2434]
### Platform

iOS

### Installed

Swift Package Manager

### Version

7.31.2

### Steps to Reproduce

1. Enable Core Data Profiling at 100% profiling rate.
2. Put a breakpoint in `-[SentryCoreDataTracker managedObjectContext:executeFetchRequest:error:originalImp:]` [on line 53](https://github.com/getsentry/sentry-cocoa/blob/f781ecba1e2ccfc37e2156f1a9300452c0c8f4c8/Sources/Sentry/SentryCoreDataTracker.m).
3. Execute a fetch request, like this:

   ```objc
   NSError *error = nil;
   [someContext executeFetchRequest:someRequest error:&error];
   ```

### Expected Result

If the fetch request succeeds, no error should be reported to Sentry.

### Actual Result

An error is reported to sentry, even though the fetch request succeeds. This is because of the following [line of code](https://github.com/getsentry/sentry-cocoa/blob/f781ecba1e2ccfc37e2156f1a9300452c0c8f4c8/Sources/Sentry/SentryCoreDataTracker.m):

```objc
[fetchSpan finishWithStatus:error != nil ? kSentrySpanStatusInternalError : kSentrySpanStatusOk];
```

This is checking whether a pointer to an `NSError *` was passed into the fetch request, not whether an error actually occurred. If I instead call my fetch request like this:

```objc
[someContext executeFetchRequest:someRequest error:nil]
                                              //   ^^^ nil here
```

Then the Sentry code will think an error did _not_ occur, even if it actually did.

The correct way to check for an error is to look at the return value of the fetch request. If it returns `nil`, then there was an error. Empty results are communicated via an empty array. I propose that the code should look like this instead:

```objc
[fetchSpan finishWithStatus:result == nil ? kSentrySpanStatusInternalError : kSentrySpanStatusOk];
```

See also #2433, where a similar problem is reported.

**Comment by brustolin** [comment]
Hey @bjhomer, thanks for pointing this out.
The actual problem is that we're not checking the content of the error pointer.

The line should be like this:
```objective-c
[fetchSpan finishWithStatus:*error != nil ? kSentrySpanStatusInternalError : kSentrySpanStatusOk];
```

**Comment by kevinrenskers** [comment]
> Hey @bjhomer, thanks for pointing this out. The actual problem is that we're not checking the content of the error pointer.
> 
> The line should be like this:
> 
> ```objc
> [fetchSpan finishWithStatus:*error != nil ? kSentrySpanStatusInternalError : kSentrySpanStatusOk];
> ```

I think @bjhomer is correct that we need to check the return value, not the error.

**Comment by brustolin** [comment]
> I think @bjhomer is correct that we need to check the return value, not the error.

He is not wrong. Is just a different way to perceive an error.

**Comment by kevinrenskers** [comment]
> **Important**: Success or failure is indicated by the return value of the method. Although Cocoa methods that indirectly return error objects in the Cocoa error domain are guaranteed to return such objects if the method indicates failure by directly returning nil or NO, you should always check that the return value is nil or NO before attempting to do anything with the NSError object.
> 
> (From the [Cocoa Error Handling Programming Guide](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ErrorHandlingCocoa/CreateCustomizeNSError/CreateCustomizeNSError.html#//apple_ref/doc/uid/TP40001806-CH204-BAJIIGCC) )

**Comment by bjhomer** [comment]
There are two problems with using `*error != nil`.

**The first problem** is that Cocoa does not promise to *clear* the error pointer if there was no error. For example, if a developer did this:

```objc
NSError *dummyError = [NSError new];
NSArray *results = [context executeFetchRequest:request error:&dummyError];
```

And if the call to `executeFetchRequest:error:` *succeeded*, then it will have had no reason to write through the error pointer, and it will likely still point to `dummyError` which is not nil. So in this case, if you check `*error != nil`, you'll incorrectly think that the call failed. The above is a contrived example, but it is _possible_, and it might be more likely to happen if the developer were reusing the same error variable across multiple calls.

**The second problem** is that the developer can pass in `nil` for the `NSError **` parameter, indicating that they don't really care about any errors, like this:

```objc
NSArray *results = [context executeFetchRequest:request error:nil];
```

If you check `*error != nil`, you'll be dereferencing a nil pointer, which will cause the process to crash. So not only is `*error != nil` the wrong thing to be checking (as per the Cocoa Error Handling Programming Guide link above), it's also *dangerous*.

**Comment by brustolin** [comment]
> Cocoa does not promise to clear the error pointer

That's is what I was looking for to understand why Cocoa Error Handling Programming Guide, suggest the way they do. Thanks.

> If you check *error != nil, you'll be dereferencing a nil pointer

I don't think this is a problem because I would use `error != nil && *error != nil`

**Comment by philipphofmann** [comment]
As we are doing a hotfix for https://github.com/getsentry/sentry-cocoa/pull/2440, we could merge the fix to both master and 8.0.0, @brustolin.
